File: apps/api/src/organizeg3_api/infrastructure/persistence/models/supplier.py
"""SQLAlchemy ORM mapping for tenant suppliers."""

from __future__ import annotations

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Index,
    Integer,
    String,
    UniqueConstraint,
    text,
)
from sqlalchemy.orm import Mapped, mapped_column

from organizeg3_api.infrastructure.database.base import (
    Base,
    TenantScopedMixin,
    TimestampMixin,
    UUIDPrimaryKeyMixin,
)


class SupplierModel(
    UUIDPrimaryKeyMixin,
    TenantScopedMixin,
    TimestampMixin,
    Base,
):
    """Represent a supplier belonging to one tenant."""

    __tablename__ = "suppliers"

    __table_args__ = (
        CheckConstraint(
            "TRIM(code) <> ''",
            name="code_not_blank",
        ),
        CheckConstraint(
            "TRIM(name) <> ''",
            name="name_not_blank",
        ),
        UniqueConstraint(
            "tenant_id",
            "code",
            name="uq_suppliers_tenant_code",
        ),
        UniqueConstraint(
            "tenant_id",
            "document_number",
            name="uq_suppliers_tenant_document_number",
        ),
        UniqueConstraint(
            "tenant_id",
            "legacy_supplier_id",
            name="uq_suppliers_tenant_legacy_supplier_id",
        ),
        Index(
            "ix_suppliers_tenant_active",
            "tenant_id",
            "is_active",
        ),
        Index(
            "ix_suppliers_tenant_name",
            "tenant_id",
            "name",
        ),
        UniqueConstraint(
            "id",
            "tenant_id",
            name="uq_suppliers_id_tenant",
        ),
    )

    legacy_supplier_id: Mapped[int | None] = mapped_column(
        Integer,
        nullable=True,
    )

    code: Mapped[str] = mapped_column(
        String(100),
        nullable=False,
    )

    name: Mapped[str] = mapped_column(
        String(255),
        nullable=False,
    )

    trade_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    legal_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    document_number: Mapped[str | None] = mapped_column(
        String(14),
        nullable=True,
    )

    state_registration: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    email: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    invoice_email: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    phone: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
    )

    secondary_phone: Mapped[str | None] = mapped_column(
        String(20),
        nullable=True,
    )

    website: Mapped[str | None] = mapped_column(
        String(500),
        nullable=True,
    )

    contact_name: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    postal_code: Mapped[str | None] = mapped_column(
        String(8),
        nullable=True,
    )

    street: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    number: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
    )

    district: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    city: Mapped[str | None] = mapped_column(
        String(255),
        nullable=True,
    )

    state: Mapped[str | None] = mapped_column(
        String(2),
        nullable=True,
    )

    is_active: Mapped[bool] = mapped_column(
        Boolean,
        nullable=False,
        default=True,
        server_default=text("true"),
    )

File: apps/pwa/src/features/suppliers/model/supplier.ts
export interface Supplier {
    readonly id: string;
    readonly tenant_id: string;
    readonly code: string;
    readonly name: string;
    readonly trade_name: string | null;
    readonly legal_name: string | null;
    readonly document_number: string | null;
    readonly state_registration: string | null;
    readonly email: string | null;
    readonly invoice_email: string | null;
    readonly phone: string | null;
    readonly secondary_phone: string | null;
    readonly website: string | null;
    readonly contact_name: string | null;
    readonly postal_code: string | null;
    readonly street: string | null;
    readonly number: string | null;
    readonly district: string | null;
    readonly city: string | null;
    readonly state: string | null;
    readonly is_active: boolean;
    readonly created_at: string;
    readonly updated_at: string;
}

export interface SupplierCreateInput {
    readonly code: string;
    readonly name: string;
    readonly trade_name?: string | null;
    readonly legal_name?: string | null;
    readonly document_number?: string | null;
    readonly state_registration?: string | null;
    readonly email?: string | null;
    readonly invoice_email?: string | null;
    readonly phone?: string | null;
    readonly secondary_phone?: string | null;
    readonly website?: string | null;
    readonly contact_name?: string | null;
    readonly postal_code?: string | null;
    readonly street?: string | null;
    readonly number?: string | null;
    readonly district?: string | null;
    readonly city?: string | null;
    readonly state?: string | null;
}

export interface SupplierUpdateInput {
    readonly code?: string;
    readonly name?: string;
    readonly trade_name?: string | null;
    readonly legal_name?: string | null;
    readonly document_number?: string | null;
    readonly state_registration?: string | null;
    readonly email?: string | null;
    readonly invoice_email?: string | null;
    readonly phone?: string | null;
    readonly secondary_phone?: string | null;
    readonly website?: string | null;
    readonly contact_name?: string | null;
    readonly postal_code?: string | null;
    readonly street?: string | null;
    readonly number?: string | null;
    readonly district?: string | null;
    readonly city?: string | null;
    readonly state?: string | null;
}

export interface SupplierListFilters {
    readonly includeInactive?: boolean;
    readonly search?: string;
    readonly limit?: number;
    readonly offset?: number;
}

export interface SupplierPage {
    readonly items: readonly Supplier[];
    readonly hasPrevious: boolean;
    readonly hasNext: boolean;
    readonly offset: number;
    readonly pageSize: number;
}

export const DEFAULT_SUPPLIER_PAGE_SIZE = 20;

export function formatSupplierDocument(
    documentNumber: string | null,
): string {
    if (documentNumber === null) {
        return "—";
    }

    const digits =
        documentNumber.replace(/\D/g, "");

    if (digits.length === 11) {
        return digits.replace(
            /^(\d{3})(\d{3})(\d{3})(\d{2})$/,
            "$1.$2.$3-$4",
        );
    }

    if (digits.length === 14) {
        return digits.replace(
            /^(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})$/,
            "$1.$2.$3/$4-$5",
        );
    }

    return documentNumber;
}

export function formatSupplierPhone(
    phone: string | null,
): string {
    if (phone === null) {
        return "—";
    }

    const digits =
        phone.replace(/\D/g, "");

    if (digits.length === 10) {
        return digits.replace(
            /^(\d{2})(\d{4})(\d{4})$/,
            "($1) $2-$3",
        );
    }

    if (digits.length === 11) {
        return digits.replace(
            /^(\d{2})(\d{5})(\d{4})$/,
            "($1) $2-$3",
        );
    }

    return phone;
}

export function formatSupplierAddress(
    supplier: Supplier,
): string {
    const streetLine = [
        supplier.street,
        supplier.number,
    ]
        .filter(
            (value): value is string =>
                value !== null &&
                value.trim().length > 0,
        )
        .join(", ");

    const cityLine = [
        supplier.district,
        supplier.city,
        supplier.state,
    ]
        .filter(
            (value): value is string =>
                value !== null &&
                value.trim().length > 0,
        )
        .join(" - ");

    const address = [
        streetLine,
        cityLine,
        supplier.postal_code,
    ]
        .filter(
            (value): value is string =>
                value !== null &&
                value.trim().length > 0,
        )
        .join(" · ");

    return address.length > 0
        ? address
        : "—";
}